File: packages/types/src/workflow/definition.ts
/**
 * Type of workflow definition
 */

import type { StaticNode } from "../node/index.js";
import type { Edge } from "./edge.js";
import type { ID, Version, Timestamp } from "../common.js";
import type { WorkflowTrigger } from "../trigger/index.js";
import type { TriggerReference } from "../trigger/template.js";
import type { VariableDefinition } from "../workflow-execution/variables.js";
import type { WorkflowConfig } from "./config.js";
import type { TriggeredSubworkflowConfig } from "./config.js";
import type { AvailableTools } from "./tool-config.js";

/**
 * Workflow Template Types
 * Used to differentiate between different types of workflows, affecting preprocessing timing and checkpointing strategies
 */
export type WorkflowTemplateType =
  /** Trigger subworkflow: must contain START_FROM_TRIGGER and CONTINUE_FROM_TRIGGER nodes, not start, end, subgraph nodes */
  | "TRIGGERED_SUBWORKFLOW"
  /** Standalone workflow: does not contain EXECUTE_TRIGGERED_SUBGRAPH triggers and does not contain SUBGRAPH nodes */
  | "STANDALONE"
  /** Dependent workflow: contains EXECUTE_TRIGGERED_SUBGRAPH trigger or SUBGRAPH node */
  | "DEPENDENT";

/**
 * Workflow Metadata Types
 * Used to store extended information
 */
export interface WorkflowMetadata {
  /** Author Information */
  author?: string;
  /** tagged array */
  tags?: string[];
  /** categorization */
  category?: string;
}

/**
 * Workflow Template Type
 *
 * Contains the basic information and structure of a workflow definition.
 * This type is used for both file-based configuration (TOML/JSON) and runtime execution.
 *
 * Key characteristics:
 * - Pure data structure with no executable functions (serializable to JSON/TOML)
 * - Includes metadata, versioning, and structural elements (nodes, edges)
 * - Used by SDK's config parser to load workflows from files
 * - Used directly by workflow engine during execution
 */
export interface WorkflowTemplate {
  /** Workflow unique identifier (required) */
  id: ID;

  /** Workflow name (required) */
  name: string;

  /** Type of workflow (e.g., 'main', 'subworkflow', 'triggered') */
  type: WorkflowTemplateType;

  /** Optional workflow description for documentation purposes */
  description?: string;

  /** Array of nodes defining all execution steps in the workflow */
  nodes: StaticNode[];

  /** Array of edges defining the connections and flow between nodes */
  edges: Edge[];

  /** Array of workflow variable definitions for declaring variables required during execution */
  variables?: VariableDefinition[];

  /** Workflow trigger definitions for declaring workflow-level event triggers */
  triggers?: (WorkflowTrigger | TriggerReference)[];

  /** Trigger subworkflow-specific configuration (only for workflows containing START_FROM_TRIGGER nodes) */
  triggeredSubworkflowConfig?: TriggeredSubworkflowConfig;

  /** Optional workflow execution behavior configuration (timeout, retry, etc.) */
  config?: WorkflowConfig;

  /** Optional metadata information (tags, author, custom properties) */
  metadata?: WorkflowMetadata;

  /** Workflow version number for tracking changes */
  version: Version;

  /** Creation timestamp (ISO 8601 format) */
  createdAt: Timestamp;

  /** Last update timestamp (ISO 8601 format) */
  updatedAt: Timestamp;

  /**
   * Available tool configurations for LLM nodes
   * 
   * Unified interface for specifying which tools are available during workflow execution.
   * Supports static initial tools and dynamic additions during execution.
   */
  availableTools?: AvailableTools;
}

File: crates/wf-types/src/workflow/definition.rs
use serde::{Deserialize, Serialize};

use super::Edge;
use super::WorkflowConfig;
use crate::hook::BaseHookConfig;
use crate::node::BaseStaticNode;
use crate::tool::AvailableTools;
use crate::workflow_execution::VariableDefinition;
use crate::Id;
use crate::Timestamp;
use crate::Version;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum WorkflowDefinitionType {
    TriggeredSubworkflow,
    Standalone,
    Dependent,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WorkflowMetadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub author: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TriggeredSubworkflowConfig {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enable_checkpoints: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_retries: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WorkflowDefinition {
    pub id: Id,
    pub name: String,
    pub description: Option<String>,
    pub r#type: Option<WorkflowDefinitionType>,
    pub version: Option<Version>,
    pub nodes: Vec<BaseStaticNode>,
    pub edges: Vec<Edge>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config: Option<WorkflowConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variables: Option<Vec<VariableDefinition>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub triggered_subworkflow_config: Option<TriggeredSubworkflowConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<WorkflowMetadata>,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub available_tools: Option<AvailableTools>,
    /// Workflow-level hooks (BEFORE_EXECUTE / AFTER_EXECUTE, per node).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hooks: Option<Vec<BaseHookConfig>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WorkflowTemplate {
    pub id: Id,
    pub name: String,
    pub description: String,
    pub definition: WorkflowDefinition,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub template_category: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub template_tags: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_public: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enabled: Option<bool>,
}
